nucypher/datastore/datastore.py

```diff
@@ -14,269 +14,189 @@
         """
-        session = session or self._session_on_init_thread
-        policy_arrangement = session.query(PolicyArrangement).filter_by(id=arrangement_id).first()
-        if not policy_arrangement:
-            raise NotFound("No PolicyArrangement {} found.".format(arrangement_id))
-        return policy_arrangement
+        Initializes a Datastore object by path.
 
-    def get_all_policy_arrangements(self, session=None) -> List[PolicyArrangement]:
+        :param db_path: Filepath to a lmdb database.
         """
-        Returns all the PolicyArrangements
+        self.db_path = db_path
+        self.__db_env = lmdb.open(db_path, map_size=self.LMDB_MAP_SIZE)
 
-        :return: The list of PolicyArrangement objects
+    @contextmanager
+    def describe(self,
+                 record_type: 'DatastoreRecord',
+                 record_id: Union[int, str],
+                 writeable: bool=False) -> 'DatastoreRecord':
         """
-        session = session or self._session_on_init_thread
-        arrangements = session.query(PolicyArrangement).all()
-        return arrangements
+        This method is used to perform CRUD operations on the datastore within
+        the safety of a context manager by returning an instance of the
+        `record_type` identified by the `record_id` provided.
 
-    def attach_kfrag_to_saved_arrangement(self, alice, id_as_hex, kfrag, session=None):
-        session = session or self._session_on_init_thread
-        policy_arrangement = session.query(PolicyArrangement).filter_by(id=id_as_hex.encode()).first()
+        When `writeable` is `False`, the record returned by this method
+        cannot be used for any operations that write to the datastore. If an
+        attempt is made to retrieve a non-existent record whilst `writeable`
+        is `False`, this method raises a `RecordNotFound` error.
 
-        if policy_arrangement is None:
-            raise NotFound("Can't attach a kfrag to non-existent Arrangement {}".format(id_as_hex))
+        When `writeable` is `True`, the record can be used to perform writes
+        on the datastore. In the event an error occurs during the write, the
+        transaction will be aborted and no data will be written, and a
+        `DatastoreTransactionError` will be raised.
 
-        if policy_arrangement.alice_verifying_key.key_data != alice.stamp:
-            raise alice.SuspiciousActivity
-
-        policy_arrangement.kfrag = bytes(kfrag)
-        self.__commit(session=session)
-
-    def del_policy_arrangement(self, arrangement_id: bytes, session=None) -> int:
-        """
-        Deletes a PolicyArrangement from the Keystore.
+        If the record is used outside the scope of the context manager, any
+        writes or reads will error.
         """
-        session = session or self._session_on_init_thread
-        deleted_records = session.query(PolicyArrangement).filter_by(id=arrangement_id).delete()
+        with suppress(ValueError):
```

Comment:
why suppress here?

Reply:
If we are given a `record_id` it can be either an `int` or a `str`. Since keys must be encoded as `bytes`, it becomes really difficult to manage both without making a mistake somewhere. So we cast all IDs to be an `int` if they are capable.

Reply:
In what situation do you expect a ValueError to be raised?

Reply:
```python
int('my_id_is_a_string')
# vs
int('123123')
```